File: 1714F.py
import sys


def run():
    n, d12, d23, d13 = map(int, input().split())
    s = d12 + d23 + d13
    x = -d12 + d13 + d23
    y = d12 - d13 + d23
    z = d12 + d13 - d23
    r = 4
    if s % 2 == 1:
        r = -1
    elif d23 == d12 + d13:
        r = 1
    elif d13 == d12 + d23:
        r = 2
    elif d12 == d23 + d13:
        r = 3
    elif x % 2 == 1 or y % 2 == 1 or z % 2 == 1 or min(x//2, y//2, z//2) <= 0 or x//2 + y//2 + z//2 > n - 1 or n == 3:
        r = -1
    if r == -1:
        print("No")
    else:
        edg = list()
        for i, v in enumerate([[0, d12, d13],
                               [d12, 0, d23],
                               [d13, d23, 0],
                               [z//2, y//2, x//2]][r - 1]):
            if i + 1 == r:
                continue
            c = r
            for _ in range(v - 1):
                edg.append((c, n))
                c = n
                n -= 1
            edg.append((c, i + 1))
        while n > max(3, r):
            edg.append((r, n))
            n -= 1
        print("Yes")
        for i in edg:
            print(*i)


def main():
    t = int(input())
    for _ in range(t):
        run()


if __name__ == "__main__":
    # with open("_input.txt", "r") as fin, open("_output.txt", "w") as fout:
    #     sys.stdin = fin
    #     sys.stdout = fout
    #     main()
    main()

File: 1714F.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

const int N = 2e5 + 10, M = 10, MOD = 1e9 + 7, HV = 151, INF = 0x3f3f3f3f;

// struct Node
// {
//     ll i, a, b;
//     Node(ll ni, ll na, ll nb)
//     {
//         this->i = ni;
//         this->a = na;
//         this->b = nb;
//     }
// };
// vector<Node> arr[N];

int ans[N];

void dfs(int m, int a, int b, int c, int n)
{
    if (n == 3)
    {
        return;
    }
    if (a == 1)
    {
        printf("%d %d\n", m, 1);
        dfs(1, a - 1, b - 1, c - 1, n);
    }
    else if (b == 1)
    {
        printf("%d %d\n", m, 2);
        dfs(2, a - 1, b - 1, c - 1, n);
    }
    else if (c == 1)
    {
        printf("%d %d\n", m, 3);
        dfs(3, a - 1, b - 1, c - 1, n);
    }
    else
    {
        printf("%d %d\n", m, n);
        dfs(n, a - 1, b - 1, c - 1, n - 1);
    }
}

void run()
{
    int n, d12, d23, d31;
    scanf("%d%d%d%d", &n, &d12, &d23, &d31);
    int mx = max({d12, d23, d31});
    // vector<int> srt = {d12, d23, d31};
    // sort(srt.begin(), srt.end());
    if (srt[0] + srt[1] != srt[2])
    {
        printf("NO\n");
        return;
    }
    printf("YES\n");
    if (mx == d12)
    {
        dfs(2, d12, 0, d23, n);
    }
    if (mx == d23)
    {
        dfs(2, d12, 0, d23, n);
    }
    if (mx == d31)
    {
        dfs(3, d31, d23, 0, n);
    }
}

int main()
{
    freopen("_input.txt", "r", stdin);
    freopen("_output.txt", "w", stdout);
    int t = 1;
    scanf("%d", &t);
    while (t--)
    // while (scanf("%d", &n), n)
    {
        run();
    }

    return 0;
}
